presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/GroupByWithConstants.java

```diff
@@ -0,0 +1,148 @@
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.google.common.collect.ImmutableList;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.SystemSessionProperties.isOptimizeConstantGroupingKeys;
+import static com.facebook.presto.sql.planner.plan.Patterns.aggregation;
+import static com.facebook.presto.sql.planner.plan.Patterns.project;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.facebook.presto.sql.relational.OriginalExpressionUtils.castToExpression;
+import static com.facebook.presto.sql.relational.OriginalExpressionUtils.isExpression;
+
+/**
+ * Transforms:
+ * <pre>
+ * - GroupBy key1, <const_expr>, key2
+ * </pre>
+ * Into:
+ * <pre>
+ * - Project <const_expr>
+ *    - GroupBy key1, key2
+ * </pre>
+ */
+public class GroupByWithConstants
+        implements Rule<AggregationNode>
+{
+    private static final Capture<ProjectNode> CHILD = Capture.newCapture();
+
+    private static final Pattern<AggregationNode> PATTERN =
+            aggregation()
+            .with(source().matching(project().capturedAs(CHILD)));
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode parent, Captures captures, Context context)
+    {
+        if (!isOptimizeConstantGroupingKeys(context.getSession())) {
+            return Result.empty();
+        }
+
+        // for each variable references in grouping keys, check if the CHILD expression defines them as constants
+        ProjectNode child = captures.get(CHILD);
+        List<VariableReferenceExpression> outputVariables = parent.getOutputVariables();
+
+        Map<VariableReferenceExpression, RowExpression> constChildVars = extractConstVars(child, outputVariables);
+        Map<VariableReferenceExpression, RowExpression> constGroupingKeys = new HashMap<>();
+
+        List<VariableReferenceExpression> groupingKeys = parent.getGroupingKeys();
+        for (VariableReferenceExpression key : groupingKeys) {
+            if (constChildVars.containsKey(key)) {
+                constGroupingKeys.put(key, constChildVars.get(key));
+            }
+        }
+
+        if (constGroupingKeys.isEmpty()) {
+            return Result.empty();
+        }
+
+        List<VariableReferenceExpression> newGroupingKeys =
+                groupingKeys.stream()
+                        .filter(key -> !constChildVars.containsKey(key))
+                        .collect(Collectors.toList());
+
+        AggregationNode newAgg = new AggregationNode(
+                parent.getSourceLocation(),
+                parent.getId(),
+                child,
+                parent.getAggregations(),
+                AggregationNode.singleGroupingSet(newGroupingKeys),
```

Comment:
In the pattern you also need to check that the aggregation only has a single grouping set.  Then you should add a test that shows the optimization is not applied if there is more than one grouping set

Reply:
done + added a new test